### ISSUE TITLE
Twitter Search Agent > Post Agent. Webhooks emitting duplicate tweets

### ISSUE THREAD

**Issue by TheChrisK** [huginn/huginn#2200]
I stood up a second server for Huginn to monitor Twitter accounts for road closures using Discord webhooks. The one environment is working fine but the new one I stood up appears to be emitting duplicate tweets to the webhooks. Sometimes the tweet is posted up to 5 times in a row. The agents are set up the same on both servers so I can't imagine it's that. The Twitter agent does not appear to be emitting the tweets more than once.

Let me know what information I can provide to help me troubleshoot this.

I am running it on Ubuntu 16.04 in Azure from source.

**Comment by dsander** [comment]
That sounds really odd, can you reproduce the problem when you manually re-emit one of the Events from the TwitterAgent?

**Comment by TheChrisK** [comment]
I can't reproduce it manually. I've tried re-emitting several events and it only sends those once.

**Comment by dsander** [comment]
Hmm if it does not happen when you re-emit the Events the Huginn instance seems to work properly. Maybe you can try and a second PostAgent that is sending the same Events to https://requestb.in/ and check if you are getting duplicates there too.

**Comment by TheChrisK** [comment]
I'll try that now and let you know.

**Comment by TheChrisK** [comment]
One thing I just noticed is the Post Agent isn't actually emitting twice. I have **emit_events** set to _true_ and the most recent duplicate does not show up twice in the events page. In my working instance, we have [this fix](https://github.com/huginn/huginn/pull/1892/commits/5c26e58cfb2b8f4d6028383e4227569315c9b401) implemented due to jobs hanging up and holding the queue but I do not have this implemented in my new instance as it did not seem to be an issue. I am still seeing errors [like this](https://pastebin.com/ebA1Fw3J) and they seem to correspond with the duplicate posts. I did try the above fix and exporting Huginn just pegs the CPU and never starts properly. I assume this was addressed in the most recent version.

**Comment by TheChrisK** [comment]
It's definitely doing it every [this error](https://pastebin.com/GBUZwTQh) comes up. It's not rapid firing the duplicates, but posting the same thing every minute or two.

**Comment by TheChrisK** [comment]
Going back to the requestb.in, the posts are being duplicated there as well.

**Comment by dsander** [comment]
The `execution expired` exception is thrown by the PostAgent that is making the requests to Discord/requestb.in? If the Agent isn't creating multiple Events when those duplicates occur it succeeds only once and even though it's execution takes too long the receiving side must have processed the request which is odd to happen so often.

If you are using the threaded background processed can you try to switch to a separate delated job process? (make sure to stop everything before modifying the `Procfile`)

```
web: bundle exec rails server -p ${PORT-3000} -b ${IP-0.0.0.0}
runner: bundle exec rails runner bin/agent_runner.rb
dj: bundle exec script/delayed_job run
```

**Comment by TheChrisK** [comment]
Getting a 502 error now on the WFE. [Here was my old Procfile config.](https://pastebin.com/fYXAK8yQ)

**Comment by dsander** [comment]
Sorry copy and pasted from the wrong section, the `web` process should stay the way it was before:
```
web: bundle exec unicorn -c config/unicorn.rb
```

**Comment by TheChrisK** [comment]
Ok, it's up again. I will monitor with this config and get back to you shortly.

**Comment by TheChrisK** [comment]
Now hung jobs are holding up the queue. If I apply the fix described [here](https://github.com/huginn/huginn/pull/1892/commits/5c26e58cfb2b8f4d6028383e4227569315c9b401), which resolved this issue previously in my other environment, the startup pegs the CPU at 100% indefinitely and the RAM wobbles back and forth.

**Comment by dsander** [comment]
The proposed fix is supposed to help preventing delayed job timeouts from messing with the other threads in the `threaded.rb` background processor. In your current setup it should not make a difference and I can not imagine why the startup would hang other than a syntax error somewhere which makes the processes reboot in a infinite loop.

How is `DELAYED_JOB_MAX_RUNTIME` configured? In my experience DJ does a good job killing jobs that run longer than the configured timeframe.

What are the differences between your new server and the old one? It is odd that the same configuration and Agents worked before.

**Comment by TheChrisK** [comment]
So far the only differences I have found are with the web_request_concern.rb (because that's where the delayed job fix is. Otherwise, the Procfile, .env and unicorn.rb are all the same.

`DELAYED_JOB_MAX_RUNTIME=1` for both environments.

**Comment by TheChrisK** [comment]
I think I may have fixed it. I copied the web_request_concern.rb from my working server to the new one and have not seen any duplicates for almost 4 hours. The errors on the post agents are more consistent with my working environment (the post times out but retries and completes w/ no duplicates). [Here is the error I am seeing now](https://pastebin.com/Y7uKphyE) which looks slightly different.

I think those extra lines in the web_request_concern.rb were conflicting with the timeout fix causing the loop. I will keep monitoring for another day to see how it goes.

**Comment by dsander** [comment]
The new error makes sense, it is thrown by the HTTP request library which is timing out after a minute. I still don't understand why the delayed job timeout did not work properly, I have never notice that before when running dedicated DJ processes. 

Is your server running out of memory or running on a really high load average?

**Comment by TheChrisK** [comment]
No resource issues that I can tell. Both VMs have 1 core and 3.5 GB of RAM. My original server has the heaviest usage and still has about 1.5 GB of RAM free and only averages about 5% CPU usage.

Of all the things I tried, I don't think I completely replaced the web_request_concern.rb from one to the other. Perhaps I did something wrong when I added that one line previously.

In any case, it's been nearly 24 hours and I have not seen any duplicates. The massive amount of tweets that come through have been timely and seemingly without issue. 

I will wait a few more hours before I officially declare victory then I will close this issue. Thanks for your attention to this. Huginn has really been a great tool for us.
